src/tokenserver/db/models.rs

```diff
@@ -1,37 +1,262 @@
-use diesel::mysql::MysqlConnection;
-use diesel::sql_types::Text;
-use diesel::{Connection, RunQueryDsl};
+use actix_web::web::block;
+#[cfg(test)]
+use diesel::sql_types::{Bigint, Integer, Nullable};
+use diesel::{
+    mysql::MysqlConnection,
+    r2d2::{ConnectionManager, PooledConnection},
+    sql_types::Text,
+    RunQueryDsl,
+};
+#[cfg(test)]
+use diesel_logger::LoggingConnection;
+use futures::future::LocalBoxFuture;
+use futures::TryFutureExt;
 
-use super::results::GetTokenserverUser;
+use std::{self, sync::Arc};
+
+use super::{params, results};
 use crate::db::error::{DbError, DbErrorKind};
+use crate::error::ApiError;
+use crate::sync_db_method;
+
+pub type DbFuture<'a, T> = LocalBoxFuture<'a, Result<T, ApiError>>;
+pub type DbResult<T> = std::result::Result<T, DbError>;
+type Conn = PooledConnection<ConnectionManager<MysqlConnection>>;
+
+#[derive(Clone)]
+pub struct TokenserverDb {
+    /// Synchronous Diesel calls are executed in actix_web::web::block to satisfy
+    /// the Db trait's asynchronous interface.
+    ///
+    /// Arc<MysqlDbInner> provides a Clone impl utilized for safely moving to
+    /// the thread pool but does not provide Send as the underlying db
+    /// conn. structs are !Sync (Arc requires both for Send). See the Send impl
+    /// below.
+    pub(super) inner: Arc<DbInner>,
+}
+
+/// Despite the db conn structs being !Sync (see Arc<MysqlDbInner> above) we
+/// don't spawn multiple MysqlDb calls at a time in the thread pool. Calls are
+/// queued to the thread pool via Futures, naturally serialized.
+unsafe impl Send for TokenserverDb {}
+
+pub struct DbInner {
+    #[cfg(not(test))]
+    pub(super) conn: Conn,
+    #[cfg(test)]
+    pub(super) conn: LoggingConnection<Conn>, // display SQL when RUST_LOG="diesel_logger=trace"
+}
+
+impl TokenserverDb {
+    pub fn new(conn: Conn) -> Self {
+        let inner = DbInner {
+            #[cfg(not(test))]
+            conn,
+            #[cfg(test)]
+            conn: LoggingConnection::new(conn),
+        };
+
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    fn get_user_sync(&self, email: String) -> DbResult<results::GetUser> {
+        let query = r#"
+            SELECT users.uid, users.email, users.client_state, users.generation,
```

Comment:
FYI: We've tried to standardize the SQL syntax in the rest of the syncstorage impl towards https://www.sqlstyle.guide/ but I won't hold you to it.

Reply:
Good to know! I'll update the syntax I added in this PR.